### PR TITLE
CMake: ensure Swift host tools depend on HostCompatibilityLibs target

### PR DESCRIPTION
**Pull request by edymtt**
At the time I implemented #60728, I did not realize that in all the bootstrapping mode we need to depend on the compatibility libraries we build -- otherwise when targeting a low deployment target we risk failing because we are not able to find such libraries.

Addresses rdar://126552386

### CODE REVIEW

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please test WebAssembly

**Comment by edymtt** [comment]
@swift-ci please test Apple Silicon

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
@swift-ci Please Test Source Compatibility Release

**Review by compnerd — APPROVED** [review]
`target_link_libraries` should almost always be universally preferred over `target_link_directories`. That ensures that the dependencies are wired up properly.

**Comment by etcwilde** [comment]
I think this changes a subtle detail here and I'm still working out whether it will still be fine or not.

These libraries are part of the toolchain content and should only be linked under certain circumstances as decided by the driver/compiler, including info about the minimum version of the runtime you're backdeploying to. That is, these should only be linked when the driver/compiler decides to do so. In the case where the driver/compiler decides to link them, it will do so automatically, so I don't think `target_link_libraries` is correct here.

Now, that said, these are static archives, so if the generated binary doesn't have one of the compatibility FORCE_LOAD symbols that it needs to resolve, the lazy-link behavior of static archives should actually skip over the entire archive since it isn't needed.

I think `add_dependencies` is probably better in this case though.

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please test WebAssembly

**Comment by edymtt** [comment]
@swift-ci please test Apple Silicon

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
@swift-ci Please Test Source Compatibility Release

**Comment by edymtt** [comment]
Updated the PR based on Evan's feedback.

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -466,6 +466,11 @@ function(_add_swift_runtime_link_flags target relpath_to_lib_dir bootstrapping)
   if(SWIFT_HOST_VARIANT_SDK IN_LIST SWIFT_DARWIN_PLATFORMS)
 
     set(sdk_dir "${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_ARCH_${SWIFT_HOST_VARIANT_ARCH}_PATH}/usr/lib/swift")
+    get_property(compatibility_libs
```
Comment (etcwilde) [inline cmake/modules/AddSwift.cmake:469]: Did `add_dependencies(${target} HostCompatibilityLibs)` not work?

**Comment by edymtt** [comment]
Source Compatibility Suite results look good -- for some reason I got only 3 failures, a subset of the 5 found in the [corresponding job](https://ci.swift.org/job/swift-main-source-compat-suite/755/)

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please test WebAssembly

**Comment by edymtt** [comment]
@swift-ci please test Apple Silicon

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
@swift-ci Please Test Source Compatibility Release

File: cmake/modules/AddSwift.cmake (hunk 1 above)
Comment (edymtt) [reply cmake/modules/AddSwift.cmake:475]: Added a comment about this in https://github.com/apple/swift/commit/4d14497016780add0b674f68a7ed4594ac8ce3a9 (as this is an aspect of interface libraries that is easy to forget)

**Pull request merged**